pkg/controller/direct/alloydb/cluster_controller.go

```diff
@@ -17,339 +17,644 @@ package alloydb
+					dayofweek.DayOfWeek_WEDNESDAY,
+					dayofweek.DayOfWeek_THURSDAY,
+					dayofweek.DayOfWeek_FRIDAY,
+					dayofweek.DayOfWeek_SATURDAY,
+					dayofweek.DayOfWeek_SUNDAY,
+				},
+				StartTimes: []*timeofday.TimeOfDay{
+					{Hours: 23},
+				},
+			},
+		}
+	}
+
+	if desired.ContinuousBackupConfig == nil {
+		desired.ContinuousBackupConfig = &alloydbpb.ContinuousBackupConfig{}
+	}
+	if desired.ContinuousBackupConfig.Enabled == nil {
+		desired.ContinuousBackupConfig.Enabled = direct.PtrTo(true)
+	}
+	if desired.ContinuousBackupConfig.RecoveryWindowDays == 0 {
+		desired.ContinuousBackupConfig.RecoveryWindowDays = 14
 	}
 
-	mc := &direct.MapContext{}
-	spec := ClusterSpecFromAPI(mc, a.actual)
-	if err := mc.Err(); err != nil {
-		return nil, fmt.Errorf("error converting alloydb cluster from API %w", err)
+	if desired.GeminiConfig == nil {
+		desired.GeminiConfig = &alloydbpb.GeminiClusterConfig{}
+	}
+	if desired.SubscriptionType == alloydbpb.SubscriptionType_SUBSCRIPTION_TYPE_UNSPECIFIED {
+		desired.SubscriptionType = alloydbpb.SubscriptionType_STANDARD
 	}
+	desired.DatabaseVersion = actual.DatabaseVersion
 
-	spec.ProjectRef.External = a.projectID
-	spec.ResourceID = &a.resourceID
+	desired.Source = actual.Source
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *ClusterAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating Cluster", "name", a.id)
+	mapCtx := &direct.MapContext{}
 
-	specObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(spec)
+	// TODO: Check immutability for optional and immutable fields.
+	// 1. Resolve reference fields.
+	if err := a.normalizeReferences(ctx); err != nil {
+		return fmt.Errorf("normalizing reference for update: %w", err)
+	}
+	// 2. Resolve secret field.
+	if err := a.resolveInitialUserField(ctx); err != nil {
+		return err
+	}
+	// 3. Set default fields that were set in the actual state.
+	a.resolveKRMDefaultsForUpdate()
+	// 4. Validate mutually-exclusive fields.
+	if a.desired.Spec.RestoreBackupSource != nil && a.desired.Spec.RestoreContinuousBackupSource != nil {
+		return fmt.Errorf("only one of 'spec.restoreBackupSource' " +
+			"and 'spec.restoreContinuousBackupSource' can be configured: " +
+			"both are configured")
+	}
+
+	desiredPb := AlloyDBClusterSpec_ToProto(mapCtx, &a.desired.DeepCopy().Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	// 5. Handle labels.
+	desiredPb.Labels = make(map[string]string)
+	for k, v := range a.desired.GetObjectMeta().GetLabels() {
+		desiredPb.Labels[k] = v
+	}
+	desiredPb.Labels["managed-by-cnrm"] = "true"
+
+	// 6. Set resource name. This step is not needed for other operations.
+	desiredPb.Name = a.id.String()
+	// 7. Handle default values for fields not yet supported in KRM types.
+	a.resolveGCPDefaults(desiredPb, a.actual)
```

Comment:
Do we need to set default values for fields that are not included in the updateMask?

Reply:
Hmmm, the fields will be included in the updateMask if we don't set default values.
We are using `CompareProtoMessage()` function to compare the diff, and any differences in between the actual proto object and the desired proto object will be reflected. However the actual state contains many fields (and they have default values) that are not supported in the KRM types. So we need to match desired state with actual state about those fields to avoid unexpected diffs.

Reply:
I guess what you asked basically is this TODO: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/7d1940e97255c6d7df7f4e23cc8418e1b47a55f1/pkg/controller/direct/alloydb/cluster_controller.go#L567?

Reply:
Oh, I see. I usually use `spec.someField != nil` to conditionally trigger the diff, which means I don't use `CompareProtoMessage()`. It turns out the amount of code is about the same, so I suppose there's no real workaround.